### DISCUSSION TITLE
GTK4-rs startup performance

### DISCUSSION THREAD

**Discussion by fragsalat** [gtk-rs/gtk4-rs#1839] [Q&A]
Hey there,

I've built an application I want to use on an raspberrypi as gui with a touchscreen. Now I got the issue that starting xserver and the application leads to a ~4 seconds long blank / black screen between the splash screen and the application. I found that the startup of GTK seem to be pretty slow. Even the simplest application (code below) takes over a second to start on a Ryzon 5 2600 (fairly old but should be fast enough to not be the bottleneck here^^). Considering my application uses CSS and XML templates for sake of improving developer experience, the startup time reduces more as it has to load more. 

I googled around could not find any GTK specific recommendation on performance improvements. As gtk is fairly old, I wonder why that is the case. So how can the performance of an gtk application be improved and the startup be reduced?

```
2024-09-18T18:35:47.763286Z  INFO gtk_test: main()
2024-09-18T18:35:47.765369Z  INFO gtk_test: run()
2024-09-18T18:35:48.461297Z  INFO gtk_test: startup
2024-09-18T18:35:48.461458Z  INFO gtk_test: activate
2024-09-18T18:35:48.970831Z  INFO gtk_test: rendered
```

```
use gtk4::glib::ExitCode;
use gtk4::prelude::{ApplicationExt, ApplicationExtManual, GtkWindowExt};
use tracing::info;
use tracing::level_filters::LevelFilter;
use tracing_subscriber::filter::Targets;
use tracing_subscriber::Layer;
use tracing_subscriber::layer::SubscriberExt;
use tracing_subscriber::util::SubscriberInitExt;


fn main() -> ExitCode {
    tracing_subscriber::registry()
        .with(
            tracing_subscriber::fmt::layer().with_filter(
                Targets::new().with_default(LevelFilter::DEBUG)
            )
        )
        .init();

    info!("main()");

    let app = gtk4::Application::builder().application_id("org.test").build();
    app.connect_startup(|app| info!("startup"));
    app.connect_activate(|app| {
        info!("activate");
        gtk4::Window::new().present();
        info!("rendered");
    });
    info!("run()");
    app.run()
}
```

**Comment by sdroege** [comment]
I would recommend running a profiler (e.g. sysprof, perf) to see what takes up that much time on your machine. I assume, `gtk4-demo` takes equally long?

**Reply by fragsalat** [reply]
Yes sadly it does. Not sure if the startup time can be influenced at all on that bare minimum example.

**Reply by sdroege** [reply]
Well, use a profiler and check where all that time is spent. For me it starts basically instantaneous :)

**Reply by fragsalat** [reply]
I did a run of a simple app with perf and found that most time is consumed by the graphics driver vc4_dri.so
I'm using  vc4-fkms-v3d driver overlay. I can't find anything related to performance improvements with that. Will look into it further. Maybe you or someone else has an idea.

![script-flamegraph](https://github.com/user-attachments/assets/c3dbce58-8965-41c5-925c-23c88c1db608)
